src/USER-HSMA/hsma2d.cpp

```diff
@@ -0,0 +1,2309 @@
+void HSMA2D::init()
+{
+	printf("Setting up HSMA2D implemented by Jiuyang Liang (Release 1.0.0)\n");
+
+	Step = 0;
+	Time = new float[2000];
+
+	Gauss = new double[Nw];
+	Weight = new double[Nw];
+	Gauss_Grule3D(Gauss, Weight, Nw);
+	S = ceil(sqrt(w));
+	double ar[S][Nw];
+	for (int tt = 0; tt < S; tt++)
+	{
+		for (int i = 0; i < Nw; i++)
+		{
+			ar[tt][i] = (((Gauss[i] + 1) / 2.0) * 2.0 - S + (tt) * 2.0) / S;
+		}
+	}
+
+	IntegralTop = new double** [S * Nw];
+	IntegralDown = new double** [S * Nw];
+	for (int i = 0; i < S * Nw; i++)
+	{
+		IntegralTop[i] = new double* [S * Nw];
+		IntegralDown[i] = new double* [S * Nw];
+	}
+	for (int i = 0; i < S * Nw; i++)
+	{
+		for (int j = 0; j < S * Nw; j++)
+		{
+			IntegralTop[i][j] = new double [4];
+			IntegralDown[i][j] = new double [4];
+		}
+	}
+	for (int m = 0; m < S; m++)
+		for (int i = 0; i < Nw; i++)
+			for (int n = 0; n < S; n++)
+				for (int j = 0; j < Nw; j++)
+				{
+					IntegralTop[m * Nw + i][n * Nw + j][0] = ar[m][i] * Lx / 2.0;
+					IntegralTop[m * Nw + i][n * Nw + j][1] = ar[n][j] * Ly / 2.0;
+					IntegralTop[m * Nw + i][n * Nw + j][2] = Lz / 2;
+					IntegralTop[m * Nw + i][n * Nw + j][3] = Weight[i] * Weight[j] * Lx * Ly / (4.0 * S * S);
+					IntegralDown[m * Nw + i][n * Nw + j][0] = ar[m][i] * Lx / 2.0;
+					IntegralDown[m * Nw + i][n * Nw + j][1] = ar[n][j] * Ly / 2.0;
+					IntegralDown[m * Nw + i][n * Nw + j][2] = -Lz / 2;
+					IntegralDown[m * Nw + i][n * Nw + j][3] = Weight[i] * Weight[j] * Lx * Ly / (4.0 * S * S);
+				}
+	
+	Rs = Lambda * sqrt((Lx / 2) * (Lx / 2) + (Ly / 2) * (Ly / 2) + (Lz / 2) * (Lz / 2));
+	PI = 3.141592653589793;
+	mu = 2.0 * PI / Lx;
+	EU = (1 - Gamma) / (1 + Gamma);
+
+	IndexReal = floor((p * p - 1) / 2.0);
+	IndexImag = ceil((p * p - 1) / 2.0);
+	NJKBound = floor(2.0 * p / 3.0);
+	NJK = (2 * NJKBound + 1) * (2 * NJKBound + 1);
+	LeftTermReal = new double* [2 * NJK];
+	LeftTermImag = new double* [2 * NJK];
+	for (int i = 0; i < 2 * NJK; i++)
+	{
+		LeftTermReal[i] = new double[IndexReal];
+		LeftTermImag[i] = new double[IndexImag];
+	}
+	for (int i = 0; i < 2 * NJK; i++)
+	{
+		for (int j = 0; j < IndexReal; j++)
+		{
+			LeftTermReal[i][j] = 0.00;
+		}
+		for (int j = 0; j < IndexImag; j++)
+		{
+			LeftTermImag[i][j] = 0.00;
+		}
+	}
+	ConstructLeftTerm((double**)LeftTermReal, (double**)LeftTermImag, (double***)IntegralTop, (double***)IntegralDown, Nw, NJKBound, p, EU, mu, Lx, Ly, Lz, S);//Construct the left term.
+
+	 //Construct Fibonacci Points And There Local/Multipole Expansion.  Fibonacci[Np][4] storage the positions and the weights.
+	Np = int(2 * F + 2.0);
+	Fibonacci = new double* [Np];
+	for (int i = 0; i < Np; i++)
+	{
+		Fibonacci[i] = new double[4];
+	}
+	QRD = new double* [Np]; QLocalRD = new double* [Np];
+	for (int i = 0; i < Np; i++)
+	{
+		QRD[i] = new double[p * p];
+		QLocalRD[i]= new double[p * p];
+	}
+	SetFibonacci(Fibonacci, F, Fp, Np, Rs, PI);
+    #pragma omp parallel
```

Comment:
All pragmas should be protected by suitable pre-processor tests. For OpenMP this would require checking for _OPENMP which is only present if the compiler supports OpenMP and the flag to enable OpenMP is used.

Reply:
Revised